pip/pip-333.md

```diff
@@ -0,0 +1,387 @@
+flowchart TB
+    Proxy.DirectProxyHandler --> Sslfactory
+    Proxy.AdminProxyHandler --> Sslfactory
+    Proxy.ServiceChannelInitializer --> Sslfactory
+    Broker.PulsarChannelInitializer --> Sslfactory
+    Client.PulsarChannelInitializer --> Sslfactory
+    Broker.WebService --> JettySSlContextFactory
+    Proxy.WebServer --> JettySSlContextFactory
+    PulsarAdmin --> AsyncHttpConnector
+    AsyncHttpConnector --> Sslfactory
+    JettySSlContextFactory --> Sslfactory
+    Client.HttpLookup.HttpClient --> Sslfactory
+    Sslfactory -.-> DefaultSslFactory
+    Sslfactory -.-> CustomSslFactory
+```
+
+# Detailed Design
+
+## Design and Implementation Details
+
+### Pulsar Common Changes
+
+A new abstract class called `SslFactory` that provides public methods to create a SSL Context, Client SSL Engine and
+Server SSL Engine. The SSL Context class returned will be of type `javax.net.ssl.SSLContext`. This class will refresh 
+the SSL Context at the configured interval. This refresh behavior is the same as the 
+current `SslContextAutoRefreshBuilder`.
+
+```java
+public abstract class SslFactory {
+    public SslFactory(long certRefreshInSec);
+    public abstract SSLEngine getClientSslEngine(String peerHost, int peerPort);
+    public abstract SSLEngine getServerSslEngine() throws GeneralSecurityException, IOException;
+    public abstract boolean needsUpdate();
+    public abstract void update() throws Exception;
+    public abstract void configure(String sslProviderString, 
+                                   Set<String> ciphers, 
+                                   Set<String> protocols, 
+                                   boolean allowInsecureConnection, 
+                                   boolean requireTrustedClientCertOnConnect, 
+                                   AuthenticationDataProvider authData, 
+                                   String tlsCustomParams);
+    // Implementation similar to SslContextAutoRefreshBuilder class which will be threadsafe
+    public SSLContext getInternalSslContext();
+}
+```
+
+A default implementation of the above SSLFactory class called `DefaultSslFactory` that will generate the SSL Context 
+and SSL Engines using File-based Certificates. It will be able to support both Java keystores and "pem/crt/key" files.
+
+```java
+public class DefaultSslFactory extends SslFactory {
+  public DefaultSslFactory(long certRefreshInSec);
+  public void configure(String sslProviderString, String keyStoreTypeString, String keyStore, String keyStorePassword,
+                        String trustStoreTypeString, String trustStore, String trustStorePassword,
+                        Set<String> ciphers, Set<String> protocols, String trustCertsFilePath,
+                        String certificateFilePath, String keyFilePath, boolean allowInsecureConnection,
+                        boolean requireTrustedClientCertOnConnect, AuthenticationDataProvider authData,
+                        boolean isKeyStoreEnabled);
```

Comment:
If PulsarSslFactory is an interface and there's a configure method that passes a value object with all possible configuration parameters, there won't be a need to have different methods for the default one.

Reply:
Made the modifications.